core/src/main/java/hudson/triggers/Trigger.java

```diff
@@ -275,7 +275,18 @@ public void run(AbstractProject p) {
                         if (t.tabs.check(cal)) {
                             LOGGER.log(Level.CONFIG, "cron triggered {0}", p);
                             try {
+                                long begin_time = System.currentTimeMillis();
                                 t.run();
+                                long end_time = System.currentTimeMillis();
+                                if ((end_time - begin_time) > CRON_THRESHOLD) {
+                                    LOGGER.log(Level.WARNING, "cron trigger {0}.run() triggered by {1} spent too "
+                                            + " much time ({2}) in its execution, other timers can be affected",
+                                            new Object[] {t.getClass().getName(), p, Util.getTimeSpanString(end_time - begin_time)});
+                                    ADMIN_MONITOR.report(t.getClass().getName(),
+                                            String.format("Trigger %s.run() triggered by %s spent too much time "
```

Comment:
Please assemble the message only once. There is no gain in using the logging late evaluation if we are doing `String.format` every time.